### PR TITLE
Remove directly accessing the m_idata and m_rdata structs internally

### PR DESCRIPTION
**Pull request by atmyers**
Apologies for the long pull request...

This is a needed step along the way to dealing with the anonymous struct issue for the id and cpu numbers.

### CODE REVIEW

File: Src/Particle/AMReX_ParticleHDF5.H [hunk 1]
```diff
@@ -740,9 +743,12 @@ ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt>
...
-                if (p.m_idata.id > 0) {
-                    for (int j = 0; j < AMREX_SPACEDIM + NStructReal; j++) {
-                        rptr[j] = p.m_rdata.arr[j];
+                if (p.id() > 0) {
+                    for (int j = 0; j < AMREX_SPACEDIM; j++) {
+                        rptr[j] = p.pos(j);
+                    }
+                    for (int j = 0; j < NStructReal; j++) {
+                        rptr[j] = p.rdata(j);
                     }
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_ParticleHDF5.H]: Doesn't seem to be right.

File: Src/Particle/AMReX_ParticleIO.H [hunk 2]
```diff
@@ -1617,29 +1617,29 @@ ParticleContainer<NStructReal,NStructInt,NArrayReal, NArrayInt>::WriteCoarsenedA
...
-                                for (int i = AMREX_SPACEDIM; i < AMREX_SPACEDIM + NStructReal; i++) {
-                                    char ws = (i == AMREX_SPACEDIM + NStructReal - 1) ? '\n' : ' ';
-                                    if (i == AMREX_SPACEDIM) {
+                                for (int i = 0; i < NStructReal; i++) {
+                                    char ws = (i == NStructReal - 1) ? '\n' : ' ';
+                                    if (i == 0) {
                                         // Multiply mass by 8 since we are only taking 1/8 of the 
                                         // total particles and want to keep the mass in the domain the same.
-                                        File << 8.0* it->m_rdata.arr[i] << ws;
+                                        File << 8.0* it->rdata(i) << ws;
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_ParticleIO.H:1639]: Looks like a lot of assumptions are made here. 3D only?  The first real being mass?

File: Src/Particle/AMReX_ParticleInit.H [hunk 3]
```diff
@@ -784,18 +783,18 @@ ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt>::InitFromBinar
...
                         if (m_verbose) {
                             amrex::AllPrint() << "BAD PARTICLE ID WOULD BE " << ParticleType::NextID() << '\n'
-                                              << "BAD PARTICLE POS " 
-                                              << AMREX_D_TERM(   p.m_rdata.pos[0],
-                                                                 << p.m_rdata.pos[2],
-                                                                 << p.m_rdata.pos[3])
+                                              << "BAD PARTICLE POS "
+                                              << AMREX_D_TERM(   p.pos(0),
+                                                              << p.pos(2),
+                                                              << p.pos(3))
```
Comment (WeiqunZhang) [inline Src/Particle/AMReX_ParticleInit.H]: 0, 2, 3?

File: Src/Particle/AMReX_ParticleIO.H (hunk 2 above)
Comment (atmyers) [reply Src/Particle/AMReX_ParticleIO.H:1639]: Yes, I agree. This function was moved from Nyx, but it's only used there and really specific to it.

File: Src/Particle/AMReX_ParticleHDF5.H (hunk 1 above)
Comment (atmyers) [reply Src/Particle/AMReX_ParticleHDF5.H]: Thanks, will fix.

File: Src/Particle/AMReX_ParticleIO.H (hunk 2 above)
Comment (atmyers) [reply Src/Particle/AMReX_ParticleIO.H:1639]: I talked to @jmsexton03 and we can remove this function now. I'll do so in a separate PR.

Comment (WeiqunZhang) [reply Src/Particle/AMReX_ParticleIO.H:1639]: Sounds good.

**Review by WeiqunZhang — APPROVED** [review]

**Pull request merged**